src/signal.cr

```diff
@@ -90,67 +25,215 @@ require "c/unistd"
 # ```
 #
 # Note:
-# - Signals are processed through the event loop and run in their own Fiber.
-# - Signals may be lost if the event loop doesn't run before exit.
 # - An uncaught exception in a signal handler is a fatal error.
-enum Signal
+enum Signal : Int32
+  HUP    = LibC::SIGHUP
+  INT    = LibC::SIGINT
+  QUIT   = LibC::SIGQUIT
+  ILL    = LibC::SIGILL
+  TRAP   = LibC::SIGTRAP
+  IOT    = LibC::SIGIOT
+  ABRT   = LibC::SIGABRT
+  FPE    = LibC::SIGFPE
+  KILL   = LibC::SIGKILL
+  BUS    = LibC::SIGBUS
+  SEGV   = LibC::SIGSEGV
+  SYS    = LibC::SIGSYS
+  PIPE   = LibC::SIGPIPE
+  ALRM   = LibC::SIGALRM
+  TERM   = LibC::SIGTERM
+  URG    = LibC::SIGURG
+  STOP   = LibC::SIGSTOP
+  TSTP   = LibC::SIGTSTP
+  CONT   = LibC::SIGCONT
+  CHLD   = LibC::SIGCHLD
+  TTIN   = LibC::SIGTTIN
+  TTOU   = LibC::SIGTTOU
+  IO     = LibC::SIGIO
+  XCPU   = LibC::SIGXCPU
+  XFSZ   = LibC::SIGXFSZ
+  VTALRM = LibC::SIGVTALRM
+  USR1   = LibC::SIGUSR1
+  USR2   = LibC::SIGUSR2
+  WINCH  = LibC::SIGWINCH
+
+  {% if flag?(:linux) %}
+    PWR    = LibC::SIGPWR
+    STKFLT = LibC::SIGSTKFLT
+    UNUSED = LibC::SIGUNUSED
+  {% end %}
+
   # Sets the handler for this signal to the passed function.
   #
   # After executing this, whenever the current process receives the
   # corresponding signal, the passed function will be run (instead of the OS
   # default).
-  def trap(block : Signal ->)
-    trap &block
-  end
-
-  # ditto
-  def trap(&block : Signal ->)
-    Event::SignalHandler.add_handler self, block
+  def trap(&handler : Signal ->) : Nil
+    Crystal::SignalHandler.trap(self, handler)
   end
 
   # Resets the handler for this signal to the OS default.
-  def reset
-    case self
-    when CHLD
-      # don't ignore by default.  send events to a waitpid service
-      trap do
-        Event::SignalChildHandler.instance.trigger
-      end
-    else
-      del_handler Proc(Int32, Void).new(Pointer(Void).new(0_u64), Pointer(Void).null)
-    end
+  def reset : Nil
+    Crystal::SignalHandler.reset(self)
   end
 
   # Clears the handler for this signal and prevents the OS default action.
-  def ignore
-    del_handler Proc(Int32, Void).new(Pointer(Void).new(1_u64), Pointer(Void).null)
+  def ignore : Nil
+    Crystal::SignalHandler.ignore(self)
   end
 
-  private def del_handler(block)
-    Event::SignalHandler.del_handler self
-    LibC.signal value, block
-  end
-
-  @@default_handlers_setup = false
+  @@setup_default_handlers = Atomic(Int32).new(0)
```

Comment:
Thoughts on an `Atomic::Once` abstraction in a later PR or do you think the `cas(0, 1)` is rare enough to not make it worth it?

Reply:
I think `CAS(0, 1)` is simple enough. I would have liked for `Atomic(Bool)` then `CAS(false, true)` to be possible, thought.

Reply:
Yeah, I think `Atomic(Bool)` would be better here.